Signal/src/ViewControllers/NewGroupViewController.m

```diff
@@ -66,243 +79,353 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder
+
+- (void)avatarTouched:(UIGestureRecognizer *)sender
+{
+    if (sender.state == UIGestureRecognizerStateRecognized) {
+        [self showChangeGroupAvatarUI:nil];
     }
 }
 
-- (BOOL)isContactHidden:(Contact *)contact
+#pragma mark - Table Contents
+
+- (void)updateTableContents
 {
-    if (contact.parsedPhoneNumbers.count < 1) {
-        // Hide contacts without any valid phone numbers.
-        return YES;
+    OWSTableContents *contents = [OWSTableContents new];
+
+    __weak NewGroupViewController *weakSelf = self;
+    ContactsViewHelper *helper = self.contactsViewHelper;
+
+    NSArray<SignalAccount *> *signalAccounts = self.contactsViewHelper.signalAccounts;
+    NSMutableSet *nonContactMemberRecipientIds = [self.memberRecipientIds mutableCopy];
+    for (SignalAccount *signalAccount in signalAccounts) {
+        [nonContactMemberRecipientIds removeObject:signalAccount.recipientId];
     }
 
-    if ([self isCurrentUserContact:contact]) {
-        // We never want to add ourselves to a group.
-        return YES;
+    // Non-contact Members
+
+    if (nonContactMemberRecipientIds.count > 0 || signalAccounts.count < 1) {
+
+        OWSTableSection *nonContactsSection = [OWSTableSection new];
+        nonContactsSection.headerTitle = NSLocalizedString(
+            @"NEW_GROUP_NON_CONTACTS_SECTION_TITLE", @"a title for the non-contacts section of the 'new group' view.");
+
+        [nonContactsSection addItem:[self createAddNonContactItem]];
+
+        for (NSString *recipientId in
+            [nonContactMemberRecipientIds.allObjects sortedArrayUsingSelector:@selector(compare:)]) {
+
+            [nonContactsSection
+                addItem:[OWSTableItem itemWithCustomCellBlock:^{
+                    NewGroupViewController *strongSelf = weakSelf;
+                    if (!strongSelf) {
+                        return (ContactTableViewCell *)nil;
+                    }
+
+                    ContactTableViewCell *cell = [ContactTableViewCell new];
+                    SignalAccount *signalAccount = [helper signalAccountForRecipientId:recipientId];
+                    BOOL isCurrentMember = [weakSelf.memberRecipientIds containsObject:recipientId];
+                    BOOL isBlocked = [helper isRecipientIdBlocked:recipientId];
+                    if (isCurrentMember) {
+                        // In the "contacts" section, we label members as such when editing an existing group.
+                        cell.accessoryMessage = NSLocalizedString(
+                            @"NEW_GROUP_MEMBER_LABEL", @"An indicator that a user is a member of the new group.");
+                    } else if (isBlocked) {
+                        cell.accessoryMessage = NSLocalizedString(
+                            @"CONTACT_CELL_IS_BLOCKED", @"An indicator that a contact has been blocked.");
+                    } else {
+                        OWSAssert(cell.accessoryMessage == nil);
+                    }
+
+                    if (signalAccount) {
```

Comment:
Great question. When I started this `ContactAccount/SignalAccount` work, I intended it be like a "contact, recipientId" tuple + metadata, i.e. that contact would always be non-nil.  After the work was mostly done I realized that it be nicer if we also use `SignalAccount` to represent non-contacts - to simplify code paths like this one specifically.  

But I'm going to do that cleanup in a subsequent PR.  It was such a struggle to cut this PR off.